apps/meteor/app/lib/server/functions/createRoom.ts

```diff
@@ -21,10 +20,90 @@ const isValidName = (name: unknown): name is string => {
 const onlyUsernames = (members: unknown): members is string[] =>
 	Array.isArray(members) && members.every((member) => typeof member === 'string');
 
+async function createUsersSubscriptions({
+	room,
+	shouldBeHandledByFederation,
+	members,
+	now,
+	owner,
+	options,
+}: {
+	room: IRoom;
+	shouldBeHandledByFederation: boolean;
+	members: string[];
+	now: Date;
+	owner: IUser;
+	options?: ICreateRoomParams['options'];
+}) {
+	if (shouldBeHandledByFederation) {
+		const extra: Partial<ISubscriptionExtraData> = options?.subscriptionExtra || {};
+		extra.open = true;
+		extra.ls = now;
+
+		if (room.prid) {
+			extra.prid = room.prid;
+		}
+
+		await Subscriptions.createWithRoomAndUser(room, owner, extra);
+
+		return;
+	}
+
+	const subs = [];
+
+	const memberIds = [];
+
+	const membersCursor = Users.findUsersByUsernames(members, {
+		projection: { 'username': 1, 'settings.preferences': 1, 'federated': 1, 'roles': 1 },
+	});
+
+	for await (const member of membersCursor) {
+		try {
+			await callbacks.run('federation.beforeAddUserToARoom', { user: member, inviter: owner }, room);
+			await callbacks.run('beforeAddedToRoom', { user: member, inviter: owner });
```

Comment:
The callback definition expects a full IUser object, but this `member` here is fetched with a projection. Maybe we can change the callback definition to specific what it actually uses?

Reply:
I improved the callback types 😉